### PR TITLE
Fix the edge case in residual stopping criterion

### PR DESCRIPTION
**Pull request by keldu**
Regarding #1051 this would fix the edge case with an rhs vector filled with zeroes.

Used '<=' instead of '<' in residual operations. Test cases will be added

### CODE REVIEW

**Comment by yhmtsai** [comment]
@keldu what's the status of this pull request? It seems to be good for review.

**Comment by greole** [comment]
Quick question, this currently stands +4000 LOC changes and 174 files changed, which seems to be to much for fixing an edge case. Could it be that it is not based on `dev`?

**Comment by keldu** [comment]
So. I'm not fully sure about the tests. I did a slightly small one, but I am wholly unsure if it even works.
What's missing anyway for now where I am stumped. Since I keep coming back without gaining any ground.

- [ ] Check all comments for the res_norm < \tau
- [ ] Check test and add proper test.

**Comment by keldu** [comment]
And I misclicked. It is not ready for review :facepalm:

**Comment by keldu** [comment]
> Quick question, this currently stands +4000 LOC changes and 174 files changed, which seems to be to much for fixing an edge case. Could it be that it is not based on `dev`?

Yeah. I rebase'd on an incorrect develop  branch. Caused a temporary issue which I didn't realised. I guess rebasing on origin/develop is the fix for that.

File: test/stop/residual_norm_kernels.cpp [hunk 1]
```diff
@@ -110,6 +110,28 @@ TYPED_TEST(ResidualNorm, CanIgorneResidualNorm)
...
+        constexpr gko::uint8 RelativeStoppingId{1};
+        bool one_changed{};
+        gko::array<gko::stopping_status> stop_status(this->ref, 1);
+        stop_status.get_data()[0].reset();
+        stop_status.set_executor(this->exec);
+
+        ASSERT_TRUE(criterion->update().residual_norm(res_norm).check(
+            RelativeStoppingId, true, &stop_status, &one_changed));
+        ASSERT_TRUE(stop_status.get_data()[0].has_converged());
+        ASSERT_TRUE(one_changed);
```
Comment (yhmtsai) [inline test/stop/residual_norm_kernels.cpp:132]:
```suggestion
            RelativeStoppingId, true, &stop_status, &one_changed));
        stop_status.set_executor(this->ref);
        ASSERT_TRUE(stop_status.get_data()[0].has_converged());
        ASSERT_TRUE(one_changed);
```
It should solve the segmentation fault

Comment (keldu) [reply test/stop/residual_norm_kernels.cpp]: I'd generally prefer if I could just setup the residual_norm to 0 and test that against the stopping status.
the whole step with the criterion seems to be unnecessary. Though I have no clue how to fix that if that is at all possible.

Comment (keldu) [reply test/stop/residual_norm_kernels.cpp]: Also should I check with rel_factory and abs_factory?

**Review by yhmtsai — DISMISSED** [review]
Miss the implicit residual test.
I think one of abs/rel should be enough? because they all go through the same kernel and the baseline calculation should be also tested in the existing tests.

**Review by MarcelKoch — APPROVED** [review]

**Comment by keldu** [comment]
I'm pretty sure my test is wrong btw. The problem mostly lies with the fact that I don't know what some of the functions do.
And the examples don't cover initalizing already solved systems by hand.

**Comment by keldu** [comment]
> Miss the implicit residual test. I think one of abs/rel should be enough? because they all go through the same kernel and the baseline calculation should be also tested in the existing tests.

I've added a rel criterion. Still not sure about my previous comment.

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 2]
```diff
@@ -52,13 +52,13 @@ namespace stop {
  * The mode for the residual norm criterion.
  *
  * - absolute:        Check for tolerance against residual norm.
- *                    $ || r || < \tau $
+ *                    $ || r || <= \tau $
```
Comment (upsj) [inline include/ginkgo/core/stop/residual_norm.hpp:55]: nit: (same below)
```suggestion
 *                    $ || r || \leq \tau $
```

**Review by upsj — CHANGES_REQUESTED** [review]
We are not testing the reference kernels yet. Otherwise LGTM!

**Comment by upsj** [comment]
@keldu A simple way to check the test does what it's supposed to to is to revert the other changes and only leave the new test in, it should fail.

**Review by upsj — APPROVED** [review]
Sorry, my mistake the reference test is included in the test for every backend.

File: test/stop/residual_norm_kernels.cpp [hunk 3]
```diff
@@ -110,6 +110,45 @@ TYPED_TEST(ResidualNorm, CanIgorneResidualNorm)
...
+TYPED_TEST(ResidualNorm, CheckIfResZeroConverges)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using NormVector = typename TestFixture::NormVector;
+    auto initial_res = gko::initialize<Mtx>({10.0}, this->exec);
+    std::shared_ptr<gko::LinOp> rhs = gko::initialize<Mtx>({10.0}, this->exec);
+    auto criterion =
+        this->factory->generate(nullptr, rhs, nullptr, initial_res.get());
+    auto rel_criterion =
+        this->rel_factory->generate(nullptr, rhs, nullptr, initial_res.get());
```
Comment (yhmtsai) [inline test/stop/residual_norm_kernels.cpp:122]: you need to set the reduction_factor to zero such that it indeed uses `=`.
For all cases with the original issue corcern,
1. absolute: reduction_factor = 0,
2. initial_resnorm: initial_res = 0 
3. rhs_norm: rhs = 0

create the new absolute, initial_resnorm, rhs_norm here not from the object in TestFixture.
You may need to add `using ValueType = T;` in class

**Review by yhmtsai — DISMISSED** [review]
The test does not reach `=` condition.
If you would also like to add the check more close to what we use in solver.
```
// it checks the rhs is zero and initial guess is the solution
std::shared_ptr<gko::LinOp> rhs = gko::initialize<Mtx>({0.0}, this->exec);
std::shared_ptr<gko::LinOp> mtx = gko::initialize<Mtx>({1.0}, this->exec);
std::shared_ptr<gko::LinOp> x = gko::initialize<Mtx>({0.0}, this->exec);
std::shared_ptr<gko::LinOp> res_norm = gko::initialize<NormVector>({0.0}, this->exec);
auto criterion =
         gko::stop::ResidualNorm<T>::build()
                      .with_reduction_factor(r<T>::value)
                      .on(this->exec)->generate(mtx, rhs, x.get(), nullptr);
// like the other test setting

ASSERT_TRUE(criterion->update().residual_norm(res_norm)check(
    RelativeStoppingId, true, &stop_status, &one_changed));
```
I think @upsj correct. you miss the reference test although they are almost identical between unified and reference in this case.
all tests also need the similar version for implicit Residual

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 4]
```diff
@@ -52,13 +52,13 @@ namespace stop {
  * The mode for the residual norm criterion.
  *
  * - absolute:        Check for tolerance against residual norm.
- *                    $ || r || < \tau $
+ *                    $ || r || \leq \tau $
  *
  * - initial_resnorm: Check for tolerance relative to the initial residual norm.
- *                    $ \frac{|| r ||}{|| r_0||} < \tau $
+ *                    $ \frac{|| r ||}{|| r_0||} <= \tau $
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm.hpp:58]:
```suggestion
 *                    $ || r || \leq \tau \times || r_0||$
```

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 5]
```diff
@@ -52,13 +52,13 @@ namespace stop {
...
  *
  * - rhs_norm:        Check for tolerance relative to the rhs norm.
- *                    $ \frac{|| r ||}{|| b ||} < \tau $
+ *                    $ \frac{|| r ||}{|| b ||} <= \tau $
```
Comment (yhmtsai) [inline include/ginkgo/core/stop/residual_norm.hpp:61]:
```suggestion
 *                    $ || r || \leq \tau \times || b ||$
```

File: test/stop/residual_norm_kernels.cpp [hunk 6]
```diff
@@ -110,6 +110,45 @@ TYPED_TEST(ResidualNorm, CanIgorneResidualNorm)
                  gko::NotSupported);
 }
 
+TYPED_TEST(ResidualNorm, CheckIfResZeroConverges)
```
[the remark this answers is not in the document]
Comment (keldu) [reply test/stop/residual_norm_kernels.cpp:113]: Ah. I missed the lower TestFixtures. Thanks for pointing it out

**Review by yhmtsai — COMMENTED** [review]
LGTM. Some compiling issues and the tests need fix.
Could you also rearrange the commit history by combining commits with the purpose? Also formatting in the end.
I do not change my reviews status yet until the tests are correct.

File: reference/test/stop/residual_norm_kernels.cpp [hunk 7]
```diff
@@ -820,6 +846,32 @@ TYPED_TEST(ImplicitResidualNorm, CanCreateFactory)
...
+    auto criterion =
+        gko::stop::ResidualNorm<T>::build().with_reduction_factor(r<T>::value);
+    .on(this->exec)->generate(mtx, rhs, x.get(), nullptr);
+    constexpr gko::uint8 RelativeStoppingId{1};
+    bool one_changed{};
+    gko::array<gko::stopping_status> stop_status(this->ref, 1);
+    stop_status.get_data()[0].reset();
+    stop_status.set_executor(this->exec);
+
+    ASSERT_TRUE(criterion->update().residual_norm(res_norm).check(
```
[the remark this answers is not in the document]
Comment (keldu) [reply reference/test/stop/residual_norm_kernels.cpp:868]: Yeah. Your're right. Will change. I just did a quick commit since I had some time.

**Comment by keldu** [comment]
!format

**Comment by keldu** [comment]
format!

File: reference/test/stop/residual_norm_kernels.cpp [hunk 8]
```diff
@@ -102,6 +102,32 @@ TYPED_TEST(ResidualNorm, CanCreateFactory)
     ASSERT_EQ(this->abs_factory_->get_executor(), this->exec_);
 }
 
+TYPED_TEST(ResidualNorm, CheckIfResZeroConverges)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using NormVector = typename TestFixture::NormVector;
+    std::shared_ptr<gko::LinOp> mtx = gko::initialize<Mtx>({1.0}, this->exec);
```
[the remark this answers is not in the document]
Comment (keldu) [reply reference/test/stop/residual_norm_kernels.cpp:110]: Why though? I mean I see it now. But that is a weird thing to have two different sets of naming schemes in the tests.

File: include/ginkgo/core/stop/residual_norm.hpp [hunk 9]
```diff
@@ -52,13 +52,13 @@ namespace stop {
  * The mode for the residual norm criterion.
  *
  * - absolute:        Check for tolerance against residual norm.
- *                    $ || r || < \tau $
+ *                    $ || r || \leq \tau $
```
[the remark this answers is not in the document]
Comment (keldu) [reply include/ginkgo/core/stop/residual_norm.hpp:55]: Just did it

File: reference/test/stop/residual_norm_kernels.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (keldu) [reply reference/test/stop/residual_norm_kernels.cpp]: I changed it. Still baffled why it's named differently in the reference.

**Comment by keldu** [comment]
> LGTM. Some compiling issues and the tests need fix. Could you also rearrange the commit history by combining commits with the purpose? Also formatting in the end. I do not change my reviews status yet until the tests are correct.

I mean I could generally squash them? It's not a big change, so I would prefer to do it that way.

**Comment by keldu** [comment]
format!

**Comment by yhmtsai** [comment]
format!

**Comment by yhmtsai** [comment]
the first commit adds the tests which are failed in pipeline https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/pipelines/1059541223 because it requires <= not <.
adding <= and updating the doc are in the following commit.

File: reference/test/CMakeLists.txt [hunk 10]
```diff
@@ -12,3 +12,6 @@ add_subdirectory(reorder)
 add_subdirectory(solver)
 add_subdirectory(stop)
 add_subdirectory(utils)
+
+
+
```
Comment (tcojean) [inline reference/test/CMakeLists.txt:17]: unneeded
```suggestion
```

File: reference/test/stop/residual_norm_kernels.cpp [hunk 11]
```diff
@@ -102,6 +103,67 @@ TYPED_TEST(ResidualNorm, CanCreateFactory)
...
+
+    {
+        auto criterion = gko::stop::ResidualNorm<T>::build()
+                             .with_reduction_factor(r<T>::value)
+                             .on(this->exec_)
+                             ->generate(mtx, rhs, x.get(), nullptr);
+        constexpr gko::uint8 RelativeStoppingId{1};
+        bool one_changed{};
+        gko::array<gko::stopping_status> stop_status(this->exec_, 1);
+        stop_status.get_data()[0].reset();
```
Comment (tcojean) [inline reference/test/stop/residual_norm_kernels.cpp:125]: not important: since this is always the same except for the baseline, maybe there is some worth in making this a function? same for the checks

**Review by tcojean — APPROVED** [review]
LGTM

File: reference/test/stop/residual_norm_kernels.cpp (hunk 11 above)
Comment (yhmtsai) [reply reference/test/stop/residual_norm_kernels.cpp]: done

**Comment by yhmtsai** [comment]
format!

**Review by MarcelKoch — APPROVED** [review]

**Review by upsj — APPROVED** [review]
